### PR TITLE
Catting file bytes

### PR DESCRIPTION
**Pull request by amiller68**
I implemented a client method that cats bytes of files using `offset` and `length` arguments. This implements more of the IPFS HTTP API specified here: https://docs.ipfs.io/reference/kubo/rpc/#api-v0-cat

Please comment on how I might better structure/organize the code so it can be included in the core repo.

### CODE REVIEW

File: ipfs-api-prelude/src/api.rs [hunk 1]
```diff
@@ -535,6 +535,41 @@ pub trait IpfsApi: Backend {
         }
     }
 
+    /// Returns the the specified bytes of an Ipfs object.
+    ///
+    /// # Examples
+    ///
+    /// ```
```
Comment (ferristseng) [inline ipfs-api-prelude/src/api.rs:542]: I believe this doctest is failing. Add `no_run` like the other examples. These are mostly just to ensure that code can compile.

**Review by ferristseng — CHANGES_REQUESTED** [review]
Thanks for your contribution! Looks great, but there were some small things to tidy up! Let me know if you have questions~

**Comment by amiller68** [comment]
I'll get on this later today. Thanks for the feedback!

File: ipfs-api-prelude/src/request/cat.rs [hunk 2]
```diff
@@ -15,6 +15,19 @@ pub struct Cat<'a> {
...
 }
 
-impl<'a> ApiRequest for Cat<'a> {
+#[derive(Serialize)]
+pub struct CatBytes<'a> {
+    #[serde(rename = "arg")]
+    pub path: &'a str,
+
+    pub offset: i64,
+    pub length: i64,
```
[the remark this answers is not in the document]
Comment (amiller68) [reply ipfs-api-prelude/src/request/cat.rs:24]: I can't speak for why the IPFS API specifies an i64; I do think that something like `usize` makes more sense, and works better with way I'm using the library. I can try and see if specifying these as a `usize` makes calls to the API mad or not.

File: ipfs-api-prelude/src/api.rs (hunk 1 above)
Comment (amiller68) [reply ipfs-api-prelude/src/api.rs]: Added.

File: ipfs-api-prelude/src/api.rs [hunk 3]
```diff
@@ -535,6 +535,41 @@ pub trait IpfsApi: Backend {
...
+    ///         .map_ok(|chunk| chunk.to_vec())
+    ///         .try_concat()
+    ///         .await;
+    ///    // Assert that we got the first 10 bytes of the file
+    ///    assert_eq!(res.unwrap().len() as i64, 10);
+    /// }
+    ///
+    /// ```
+
+    fn cat_bytes(&self, path: &str, offset: i64, length: i64) -> BoxStream<Bytes, Self::Error> {
```
[the remark this answers is not in the document]
Comment (amiller68) [reply ipfs-api-prelude/src/api.rs]: I went with cat_range

File: ipfs-api-prelude/src/request/cat.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (amiller68) [reply ipfs-api-prelude/src/request/cat.rs]: usize works

**Comment by amiller68** [comment]
@ferristseng I think I addressed your concerns in my latest commit. Please let me know if this is more up-to-snuff

**Comment by ferristseng** [comment]
> @ferristseng I think I addressed your concerns in my latest commit. Please let me know if this is more up-to-snuff

Looks great! Thank you!

**Pull request merged**